src/lib/libutil/efun.3
.\"     $NetBSD: efun.3,v 1.11 2015/01/18 18:09:36 christos Exp $
.\"
.\" Copyright (c) 2006 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Christos Zoulas.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd May 3, 2010
.Dt EFUN 3
.Os
.Sh NAME
.Nm esetfunc ,
.Nm easprintf ,
.Nm efopen ,
.Nm emalloc ,
.Nm ecalloc ,
.Nm erealloc ,
.Nm ereallocarr ,
.Nm estrdup ,
.Nm estrndup ,
.Nm estrlcat ,
.Nm estrlcpy ,
.Nm estrtoi ,
.Nm estrtou ,
.Nm evasprintf
.Nd error-checked utility functions
.Sh LIBRARY
.Lb libutil
.Sh SYNOPSIS
.In util.h
.Ft void (*)(int, const char *, ...)
.Fn esetfunc "void (*)(int, const char *, ...)"
.Ft int
.Fn easprintf "char ** restrict str" "const char * restrict fmt" "..."
.Ft FILE *
.Fn efopen "const char *p" "const char *m"
.Ft void *
.Fn ecalloc "size_t n" "size_t c"
.Ft void *
.Fn emalloc "size_t n"
.Ft void *
.Fn erealloc "void *p" "size_t n"
.Ft void
.Fn ereallocarr "void *p" "size_t n" "size_t n"
.Ft char *
.Fn estrdup "const char *s"
.Ft char *
.Fn estrndup "const char *s" "size_t len"
.Ft size_t
.Fn estrlcat "char *dst" "const char *src" "size_t len"
.Ft size_t
.Fn estrlcpy "char *dst" "const char *src" "size_t len"
.Ft intmax_t
.Fn estrtoi "const char * nptr" "int base" "intmax_t lo" "intmax_t hi"
.Ft uintmax_t
.Fn estrtou "const char * nptr" "int base" "uintmax_t lo" "uintmax_t hi"
.Ft int
.Fn evasprintf "char ** restrict str" "const char * restrict fmt" "..."
.Sh DESCRIPTION
The
.Fn easprintf ,
.Fn efopen ,
.Fn ecalloc ,
.Fn emalloc ,
.Fn erealloc ,
.Fn ereallocarr ,
.Fn estrdup ,
.Fn estrndup ,
.Fn estrlcat ,
.Fn estrlcpy ,
.Fn estrtoi ,
.Fn estrtou ,
and
.Fn evasprintf
functions
operate exactly as the corresponding functions that do not start with an
.Sq e
except that in case of an error, they call
the installed error handler that can be configured with
.Fn esetfunc .
.Pp
For the string handling functions, it is an error when the destination
buffer is not large enough to hold the complete string.
For functions that allocate memory or open a file, it is an error when
they would return a null pointer.
The default error handler is
.Xr err 3 .
The function
.Fn esetfunc
returns the previous error handler function.
A
.Dv NULL
error handler will just call
.Xr exit 3 .
.Sh SEE ALSO
.Xr asprintf 3 ,
.Xr calloc 3 ,
.Xr err 3 ,
.Xr exit 3 ,
.Xr fopen 3 ,
.Xr malloc 3 ,
.Xr realloc 3 ,
.Xr reallocarr 3 ,
.Xr strdup 3 ,
.Xr strlcat 3 ,
.Xr strlcpy 3 ,
.Xr strndup 3 ,
.Xr strtoi 3 ,
.Xr strtou 3 ,
.Xr vasprintf 3
.Sh HISTORY
The
.Fn estrtoi ,
.Fn estrtou ,
and
.Fn ereallocarr
functions were added in
.Nx 8 .


src/lib/libutil/efun.c
/*	$NetBSD: efun.c,v 1.9 2015/01/18 18:09:36 christos Exp $	*/

/*-
 * Copyright (c) 2006 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Christos Zoulas.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#if HAVE_NBTOOL_CONFIG_H
#include "nbtool_config.h"
#endif

#include <sys/cdefs.h>
#ifdef __RCSID
__RCSID("$NetBSD: efun.c,v 1.9 2015/01/18 18:09:36 christos Exp $");
#endif

#include <err.h>
#include <errno.h>
#include <inttypes.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <util.h>

static void (*efunc)(int, const char *, ...) = err;

void (*
esetfunc(void (*ef)(int, const char *, ...)))(int, const char *, ...)
{
	void (*of)(int, const char *, ...) = efunc;
	efunc = ef == NULL ? (void (*)(int, const char *, ...))exit : ef;
	return of;
}

size_t
estrlcpy(char *dst, const char *src, size_t len)
{
	size_t rv;
	if ((rv = strlcpy(dst, src, len)) >= len) {
		errno = ENAMETOOLONG;
		(*efunc)(1,
		    "Cannot copy string; %zu chars needed %zu provided",
		    rv, len);
	}
	return rv;
}

size_t
estrlcat(char *dst, const char *src, size_t len)
{
	size_t rv;
	if ((rv = strlcat(dst, src, len)) >= len) {
		errno = ENAMETOOLONG;
		(*efunc)(1,
		    "Cannot append to string; %zu chars needed %zu provided",
		    rv, len);
	}
	return rv;
}

char *
estrdup(const char *s)
{
	char *d = strdup(s);
	if (d == NULL)
		(*efunc)(1, "Cannot copy string");
	return d;
}

char *
estrndup(const char *s, size_t len)
{
	char *d = strndup(s, len);
	if (d == NULL)
		(*efunc)(1, "Cannot copy string");
	return d;
}

void *
emalloc(size_t n)
{
	void *p = malloc(n);
	if (p == NULL && n != 0)
		(*efunc)(1, "Cannot allocate %zu bytes", n);
	return p;
}

void *
ecalloc(size_t n, size_t s)
{
	void *p = calloc(n, s);
	if (p == NULL && n != 0 && s != 0)
		(*efunc)(1, "Cannot allocate %zu blocks of size %zu", n, s);
	return p;
}

void *
erealloc(void *p, size_t n)
{
	void *q = realloc(p, n);
	if (q == NULL && n != 0)
		(*efunc)(1, "Cannot re-allocate %zu bytes", n);
	return q;
}

void
ereallocarr(void *p, size_t n, size_t size)
{
	int rv = reallocarr(p, n, size);
	if (rv != 0) {
		errno = rv;
		(*efunc)(1, "Cannot re-allocate %zu * %zu bytes", n, size);
	}
}

FILE *
efopen(const char *p, const char *m)
{
	FILE *fp = fopen(p, m);
	if (fp == NULL)
		(*efunc)(1, "Cannot open `%s'", p);
	return fp;
}

int
easprintf(char ** __restrict ret, const char * __restrict format, ...)
{
	int rv;
	va_list ap;
	va_start(ap, format);
	if ((rv = vasprintf(ret, format, ap)) == -1)
		(*efunc)(1, "Cannot format string");
	va_end(ap);
	return rv;
}

int
evasprintf(char ** __restrict ret, const char * __restrict format, va_list ap)
{
	int rv;
	if ((rv = vasprintf(ret, format, ap)) == -1)
		(*efunc)(1, "Cannot format string");
	return rv;
}

intmax_t
estrtoi(const char * nptr, int base, intmax_t lo, intmax_t hi)
{
	int e;
	intmax_t rv = strtoi(nptr, NULL, base, lo, hi, &e);
	if (e != 0) {
		errno = e;
		(*efunc)(1,
		    "Cannot convert string value '%s' with base %d to a number in range [%jd .. %jd]",
		    nptr, base, lo, hi);
	}
	return rv;
}

uintmax_t
estrtou(const char * nptr, int base, uintmax_t lo, uintmax_t hi)
{
	int e;
	uintmax_t rv = strtou(nptr, NULL, base, lo, hi, &e);
	if (e != 0) {
		errno = e;
		(*efunc)(1,
		    "Cannot convert string value '%s' with base %d to a number in range [%ju .. %ju]",
		    nptr, base, lo, hi);
	}
	return rv;
}
